### PR TITLE
Bump kubernetes/client-go

### LINKED ISSUE #2840: revision of client-go dependency in vendor needs bumped!
### Do you want to request a *feature* or report a *bug*?
feature

### What did you expect to see?
[https://github.com/containous/traefik/blob/6a92ac0b7b04ea69a995f7635c0a7fb81fdfc952/Gopkg.toml#L189](https://github.com/containous/traefik/blob/6a92ac0b7b04ea69a995f7635c0a7fb81fdfc952/Gopkg.toml#L189)
I think it's sort of outdated. Any plan on upgrading this?

And kubernetes release-1.5 seems not maintained any more. Bumping this can activate more client-side utilities without breaking the compatibility. WDYT?

| Branch         | Canonical source code location       | Maintenance status            |
|----------------|--------------------------------------|-------------------------------|
| client-go 1.4  | Kubernetes main repo, 1.4 branch     | = -                           |
| client-go 1.5  | Kubernetes main repo, 1.5 branch     | = -                           |
| client-go 2.0  | Kubernetes main repo, 1.5 branch     | = -                           |
| client-go 3.0  | Kubernetes main repo, 1.6 branch     | = -                           |
| client-go 4.0  | Kubernetes main repo, 1.7 branch     | ✓                             |
| client-go 5.0  | Kubernetes main repo, 1.8 branch     | ✓                             |
| client-go 6.0  | Kubernetes main repo, 1.9 branch     | ✓                             |
| client-go HEAD | Kubernetes main repo, master branch  | ✓                             |

### PR DESCRIPTION
**Pull request by yue9944882**
### What does this PR do?

Briefly, this PR does the following things:
1. [x] Edit `Gopkg.toml`: Constraints client-go version to v6 and kubernetes/apixxx to release-1.9 branch.
2. [x] Align those moved / renamed packages in kubernetes repo.
3. [x] Strict kubernetes package naming: v1 -> corev1, v1beta1 -> extensionsv1beta1
4. [x] Fixes removed/deprecated functions since cilent-go 2.0.0.

### Motivation

Fixes #2840.

### More

- [x] Added/updated tests

### CODE REVIEW

**Comment by ldez** [comment]
@yue9944882 we don't really like WIP PR because:
- each time you commit, you take 2 executors in our CI (we have only 4)
- can block some other changes
- we are notify of each commit

Now, you have done this PR, could you work in another branch and you are ready push on this branch.

And could you rebase your PR on master.

**Comment by ldez** [comment]
@yue9944882 please rebase on master and use `make dep-ensure`

https://github.com/containous/traefik/blob/master/CONTRIBUTING.md#setting-up-dependency-management

**Comment by yue9944882** [comment]
@ldez Okay, I will only push next time when this PR is completely ready. (Facepalm

**Comment by timoreimann** [comment]
@yue9944882 Each client-go version should be completely backwards compatible, so I suggest we skip version 5 and go straight for version 6.

**Comment by timoreimann** [comment]
In case you missed it: The [official blog post on client-go v6](http://blog.kubernetes.io/2018/01/introducing-client-go-version-6.html) has some details on how to configure dep for v6 properly.

**Comment by yue9944882** [comment]
@timoreimann @ldez Even though this PR looks extremely large. But actually the changes is small. I will point out these changes in Code Review.

File: Gopkg.toml [hunk 1]
```diff
@@ -190,7 +190,19 @@
 
 [[constraint]]
   name = "k8s.io/client-go"
-  version = "2.0.0"
+  version = "6.0.0"
```
Comment (yue9944882) [inline Gopkg.toml]: @timoreimann Bumps happens here.

File: provider/kubernetes/client.go [hunk 2]
```diff
@@ -285,37 +284,11 @@ func (c *clientImpl) GetSecret(namespace, name string) (*v1.Secret, bool, error)
...
 
-// newListWatchFromClientWithLabelSelector creates a new ListWatch from the given parameters.
-// It extends cache.NewListWatchFromClient to support label selectors.
-func newListWatchFromClientWithLabelSelector(c cache.Getter, resource string, namespace string, fieldSelector fields.Selector, labelSelector labels.Selector) *cache.ListWatch {
-	listFunc := func(options api.ListOptions) (runtime.Object, error) {
-		return c.Get().
-			Namespace(namespace).
-			Resource(resource).
-			VersionedParams(&options, api.ParameterCodec).
-			FieldsSelectorParam(fieldSelector).
```
Comment (yue9944882) [inline provider/kubernetes/client.go]: This functions `FieldsSelectorParam` and `LabelsSelectorParam ` is removed.

File: provider/kubernetes/client.go [hunk 3]
```diff
@@ -220,22 +224,17 @@ func loadInformer(listWatch cache.ListerWatcher, object runtime.Object, watchCh
 		resyncPeriod,
 	)
 
-	if err := informer.AddEventHandler(newResourceEventHandler(watchCh)); err != nil {
-		// This should only ever fail if we add an event handler after the
-		// informer has been started already, which would be a programming bug.
-		panic(err)
-	}
-
+	informer.AddEventHandler(newResourceEventHandler(watchCh))
```
Comment (yue9944882) [inline provider/kubernetes/client.go]: `AddEventHandler ` has no return value now.

File: provider/kubernetes/client.go [hunk 4]
```diff
@@ -188,14 +188,18 @@ func (c *clientImpl) WatchAll(namespaces Namespaces, labelSelector string, stopC
...
 // WatchIngresses sets up a watch on Ingress objects and returns a corresponding shared informer.
 func (c *clientImpl) WatchIngresses(namespace string, labelSelector labels.Selector, watchCh chan<- interface{}) cache.SharedInformer {
-	listWatch := newListWatchFromClientWithLabelSelector(
-		c.clientset.ExtensionsV1beta1().RESTClient(),
-		kindIngresses,
-		namespace,
-		fields.Everything(),
-		labelSelector)
-
-	informer := loadInformer(listWatch, &v1beta1.Ingress{}, watchCh)
```
Comment (yue9944882) [inline provider/kubernetes/client.go]: ListerWatcher is moved from L301 to here.

**Comment by ldez** [comment]
@yue9944882 Could you revert all the changes on the vendor and leave the lock and the toml file, as first step.

**Comment by ldez** [comment]
If not possible please use `make dep-ensure` before push.

**Comment by ldez** [comment]
For the continuation, make commits like that.:
1. one commit for `Gopkg.toml` and `Gopkg.lock`
2. one commit for the Træfik code
3. one commit for the vendored files (after the review)

Please respect this order, in a first time please don't commit vendored files.

**Comment by ldez** [comment]
> Actually reverting vendor will fail this PR. 

I know, but for now it's not important. Please revert.

You need to use dep 0.4.1 minimum, please check you version: `dep version`.

**Comment by ldez** [comment]
> I think this can a potential BUG of dep.

And it's not a bug, it's the behavior of dep when a dependency don't use dep.

**Comment by yue9944882** [comment]
> git reset 7d3dd5a0e43b701a8231a8eff8d8024fcc7a5dba
git add Gopkg.*
git commit -m "chore: update dep configuration."
git add provider/ integration/
git commit -m "feat: update Traefik code."

@ldez Oops, I didn't notice these edition of the message you leave. How about editing these commit messages altogether after review?

**Comment by yue9944882** [comment]
@ldez @timoreimann PTAL. The tests failing tho because missing upgrades of `vendor`.

**Comment by ldez** [comment]
Not needed 😉

**Comment by ldez** [comment]
Please wait, we need time to review.
For us, it's the night 😉

File: Gopkg.toml [hunk 5]
```diff
@@ -190,7 +190,19 @@
...
+
+[[constraint]]
+  name = "k8s.io/api"
+  version = "kubernetes-1.9.0"
+
+[[constraint]]
+  name = "k8s.io/apimachinery"
+  version = "kubernetes-1.9.0"
+
+[[constraint]]
```
[the remark this answers is not in the document]
Comment (timoreimann) [reply Gopkg.toml]: It's more likely that we don't need it:  `k8s.io/apiextensions-apiserver` is only required when dealing with things like CRDs, which we don't.

If not used directly, I'd recommend to remove the stanza.

**Comment by yue9944882** [comment]
@timoreimann I just find that `k8s.io/kube-openapi` is a transitive dependency by `k8s.io/apimachinery`. Adding this to constraints.

**Comment by ldez** [comment]
@yue9944882 it's good to me, now you can add the last commit with the vendor.

**Pull request merged**